grout/models.py

```diff
@@ -52,43 +75,138 @@ def validate_schema(self, schema):
 
-    schema = models.ForeignKey('RecordSchema', on_delete=models.CASCADE)
-    data = JSONField()
+    class Meta(object):
+        abstract = True
 
-    archived = models.BooleanField(default=False)
+
+class PolygonFields(models.Model):
+    """
+    Base class for a record with polygon geometry.
+    """
+    class GeometryTypes(object):
+        POLYGON = 'polygon'
+        CHOICES = ((POLYGON, 'Polygon'),)
+
+    geom = models.PolygonField(srid=settings.GROUT['SRID'])
+    location_text = models.CharField(max_length=200, null=True, blank=True)
+    geometry_type = models.CharField(max_length=7,
+                                     choices=GeometryTypes.CHOICES,
+                                     default=GeometryTypes.POLYGON)
 
+    class Meta(object):
+        abstract = True
+
+
+class NoGeometryFields(models.Model):
+    """
+    Base class for a record with no geometry.
+    """
+    class GeometryTypes(object):
+        NONE = 'none'
+        CHOICES = ((NONE, 'None'),)
+
+    geometry_type = models.CharField(max_length=7,
+                                     choices=GeometryTypes.CHOICES,
+                                     default=GeometryTypes.NONE)
+
+    class Meta(object):
+        abstract = True
+
+
+class FlexibleRecord(AbstractFlexibleRecord, NoGeometryFields):
+    """
+    Catalog data with a flexible schema.
+    """
     class Meta(object):
         ordering = ('-created',)
 
 
-class RecordType(GroutModel):
-    """ Store extra information for a given RecordType, associated schemas in RecordSchema """
-    label = models.CharField(max_length=64)
-    plural_label = models.CharField(max_length=64)
-    description = models.TextField(blank=True, null=True)
-    active = models.BooleanField(default=True)
+class TemporalFlexibleRecord(AbstractFlexibleRecord, NoGeometryFields, DateTimeRange):
+    """
+    Catalog data with a flexible schema, including enforced date/time data.
+    """
+    class Meta(object):
+        ordering = ('-created',)
 
-    def get_current_schema(self):
-        schemas = self.schemas.order_by('-version')
-        return schemas[0] if len(schemas) > 0 else None
+
+class PointRecord(AbstractFlexibleRecord, PointFields):
+    """
+    Catalog a point in space.
+    """
+    class Meta(object):
+        ordering = ('-created',)
 
 
-class RecordSchema(SchemaModel):
-    """Schemas for spatiotemporal records"""
-    record_type = models.ForeignKey('RecordType',
-                                    related_name='schemas',
-                                    on_delete=models.CASCADE)
+class Record(AbstractFlexibleRecord, PointFields, DateTimeRange):
+    """
+    Catalog a point in time and space.
 
+    The name is perhaps confusing, but is here for legacy support. Should be
+    thought of as 'TemporalPointRecord' instead.
+    """
     class Meta(object):
-        unique_together = (('record_type', 'version'),)
+        ordering = ('-created',)
+
+
+class TemporalPointRecord(Record):
```

Comment:
The main reason I aliased `TemporalPointRecord` to `Record` instead of doing it the other way around is that I didn't want [multi-table inheritance](https://docs.djangoproject.com/en/2.0/topics/db/models/#multi-table-inheritance) to mean that the only data stored in the `Record` table was a reference to the `TemporalPointRecord` table. My motivation here was somewhat lazy: I just didn't want to write a data migration to turn all `Record`s into `TemporalPointRecord`s and then implement the references on `Record`. It seemed like a more incremental change to add a new model that would reference the old `Record` model instead.

All that being said, it doesn't seem reasonable now that I write it out like that, so I'll do the slightly harder/more correct thing and implement the data model correctly. `Record = TemporalPointRecord` is an interesting approach, and might get around some of these migration issues anyway.